src/main/scala/org/broadinstitute/dsde/firecloud/EntityClient.scala

```diff
@@ -111,5 +119,171 @@ class EntityClient (requestContext: RequestContext) extends Actor {
+    val entitiesToUpdate = tsv.tsvData.map(_.headOption.get)
+    val distinctEntities = entitiesToUpdate.distinct
+    if ( entitiesToUpdate.size != distinctEntities.size ) {
+      requestContext.complete( HttpResponse(BadRequest, "TSV has duplicated entities: " + entitiesToUpdate.diff(distinctEntities).distinct.mkString(", ")) )
+    } else {
+      op
+    }
+  }
+
+  //Bail with a 400 Bad Request if the tsv is trying to set members on a collection type.
+  //Otherwise, carry on.
+  private def checkNoCollectionMemberAttribute( tsv: TSVLoadFile, memberTypeOpt: Option[String] )(op: => Unit): Unit = {
+    if( memberTypeOpt.isDefined && tsv.headers.contains(memberTypeOpt.get + "_id") ) {
+      requestContext.complete( HttpResponse(BadRequest, "Can't set collection members along with other attributes; please use two-column TSV format or remove " + memberTypeOpt.get + "_id from your tsv.") )
+    } else {
+      op
+    }
+  }
+
+  //Verifies that the provided list of headers includes all attributes required by the schema for this entity type.
+  //Bails with a 400 Bad Request if the entity type is unknown or if some attributes are missing.
+  //Returns the list of required attributes if all is well.
+  private def withRequiredAttributes(entityType: String, headers: Seq[String])(op: (Map[String, String] => Unit)):Unit = {
+    ModelSchema.getRequiredAttributes(entityType) match {
+      case Failure(regret) => requestContext.complete(HttpResponse(BadRequest, regret.getMessage))
+      case Success(requiredAttributes) => {
+        if( !requiredAttributes.keySet.subsetOf(headers.toSet) ) {
+          requestContext.complete( HttpResponse(BadRequest, "TSV is missing required attributes: " + (requiredAttributes.keySet -- headers).mkString(", ")) )
+        } else {
+          op(requiredAttributes)
+        }
+      }
+    }
+  }
+
+  def batchCallToRawls( workspaceNamespace: String, workspaceName: String, calls: Seq[EntityUpdateDefinition] ) = {
+    log.info("TSV upload request received")
+    val pipeline: HttpRequest => Future[HttpResponse] =
+      addHeader(Cookie(requestContext.request.cookies)) ~> sendReceive
+    val responseFuture: Future[HttpResponse] = pipeline {
+      Post(FireCloudConfig.Workspace.entityPathFromWorkspace(workspaceNamespace, workspaceName)+"/batchUpsert",
+            HttpEntity(MediaTypes.`application/json`,calls.toJson.toString))
+    }
+
+    responseFuture onComplete {
+      case Success(response) =>
+        response.status match {
+          case NoContent =>
+            log.debug("OK response")
+            requestContext.complete(OK)
+          case _ =>
+            // Bubble up all other unmarshallable responses
+            log.warning("Unanticipated response: " + response.status.defaultMessage)
+            requestContext.complete(response)
+        }
+      case Failure(error) =>
+        // Failure accessing service
+        log.error(error, "Service API call failed")
+        requestContext.failWith(
```

Comment:
does failWith do the desired thing here? IIRC it does not and you really want to call complete here.